pkg/controller/perconaservermongodb/finalizers.go

```diff
@@ -46,35 +45,29 @@ func (r *ReconcilePerconaServerMongoDB) checkFinalizers(ctx context.Context, cr
 			case errWaitingTermination:
 			default:
 				log.Error(err, "failed to run finalizer", "finalizer", f)
+				if cr.Status.State == api.AppStateError {
+					continue
+				}
 			}
 			finalizers = append(finalizers, orderedFinalizers[i:]...)
 			break
 		}
 	}
 
 	cr.SetFinalizers(finalizers)
-	err = r.client.Update(ctx, cr)
+	if err := r.client.Update(ctx, cr); err != nil {
+		return false, errors.Wrap(err, "update")
+	}
 
-	return shouldReconcile, err
+	return shouldReconcile, nil
 }
 
-func (r *ReconcilePerconaServerMongoDB) deletePSMDBPods(ctx context.Context, cr *api.PerconaServerMongoDB) (err error) {
+func (r *ReconcilePerconaServerMongoDB) deletePSMDBPods(ctx context.Context, cr *api.PerconaServerMongoDB) error {
 	log := logf.FromContext(ctx)
 
 	if cr.Spec.Sharding.Enabled {
 		cr.Spec.Sharding.Mongos.Size = 0
 
-		sts := new(appsv1.StatefulSet)
-		err := r.client.Get(ctx, cr.MongosNamespacedName(), sts)
-		if client.IgnoreNotFound(err) != nil {
-			return errors.Wrap(err, "failed to get mongos statefulset")
-		}
-		if sts.Spec.Replicas != nil && *sts.Spec.Replicas > 0 {
-			err = r.disableBalancer(ctx, cr)
-			if err != nil {
-				return errors.Wrap(err, "failed to disable balancer")
-			}
-		}
```

Comment:
I removed this code, because `disableBalancer` doesn't do anything if `cr.Spec.Sharding.Mongos.Size == 0.`